### DISCUSSION TITLE
Multiple Apps

### DISCUSSION THREAD

**Discussion by JPHutchins** [beeware/briefcase#1888] [Q&A]
I decided to convert a project to multiple apps in order to share some GUI base classes.  It works great!

When it comes to the pyproject.toml, I found that I had to copy out all of the entries like `[tool.briefcase.app.firstapp.linux.system.suse]` and repeat them with `[tool.briefcase.app.secondapp.linux.system.suse]` etc.  Is there anyway to avoid this repetition and declare that all of the apps have the same briefcase dependencies?

**Comment by freakboy3742** [comment]
If the dependency is common to all app *on all platforms*, you can include the requirement in the project-level `requires` setting (i.e. in `[tool.briefcase]`)

If the dependency is linux specific, but shouldn't be used on macOS/Windows, there's no "across all apps, on linux" section where you can put a declaration. That might be worth exploring as a feature; the question would be the exact resolution order (i.e., do platform level declarations come before or after app level declarations?)

However, there is another option: you *can* use PEP508 install qualifiers in the app-level section. For example:

```
[tool.briefcase]
...
requires = [
    "pillow==9.3.0; sys_platform == 'linux'",
    "pillow==9.2.0; sys_platform == 'darwin'",
]
```
This would install Pillow 9.3.0 on Linux, but Pillow 9.2.0 on macOS. You can use any of the qualifiers here (so, for example, you can gate installation on the Python version to install a backwards compatibility shim like `importlib_metadata`).

The only complication with this is that `sys.platform == 'linux'` on Android as well (at least, it does until Python 3.13; as of Python 3.13, Android reports as `sys.platform == "android"`). But if you're not planning to target Android, that's not a concern, it's entirely safe to use.

**Comment by chenkianwee** [comment]
Hi I am trying to create a multi app project too. My main motivation for doing that is both apps share many functions I have written. I am hoping by putting them in a single project I can reference the same module for both apps. However, I am not sure where to put the shared library in the briefcase folder structure. I tried doing this:

```
Myapp
     |----- src
          |----- myapp1
          |----- myapp2
          |----- shared_libs
     |----- tests
```
In my code myapp1/app.py I tried importing shared libs as so 

`from ..shared_libs import myshared_libs`

But I am getting this error when i run briefcase dev 
'ImportError: attempted relative import beyond top-level package'

what is the correct way to share a module between two apps in a single project? Thanks

**Reply by freakboy3742** [reply]
The rules about sharing code aren't any different than they are for a "normal" Python app - and as with a normal Python app, the key is ensuring that the code is on your PYTHONPATH where you expect.

The example directory layout you've shown is entirely fine as an input - but that's the point - it's *just* the input. That's not necessarily how the directory looks when it's deployed at runtime.

By default, your app `My App 1` will have a `sources` definition of `sources = ["src/myapp1"]` or similar. That means that the `myapp1` folder will be copied into your production app. `myapp2` and `shared_libs` won't be. 

If you add `shared_libs` to your sources definition (`sources = ["src/myapp1", "src/shared_libs"]`), the `shared_libs` directory will also be copied into your final app - but there won't be a `..` namespace to traverse. `..` isn't traversing *directories*, it's traversing the Python namespace - and `shared_libs` is at the top of the Python namespace tree. There is no module "above" `myapp1` - `myapp1` is a "top-level package", available directly on your `PYTHONPATH`.

However - you *could* run `from shared_libs import ...` - because `shared_libs` is *also* at the top of the Python namespace tree.

I hope that clarifies things; let me know if you need any further clarification.

**Comment by chenkianwee** [comment]
Thanks so much! It worked! I included "src/shared_libs" as one of my sources in the pyproject.toml file. Now I can import the library in my app.

**Reply by chenkianwee** [reply]
Might be useful for others who might be looking for managing multiple apps projects. This is how I setup testing for my different apps, will be good to hear if this is the recommended practice.
```
    proj_folder
        |--- src
        |--- tests
            |--- app1
                |--- test_app.py
            |--- app2
                |--- test_app.py
            |--- app1.py
            |--- app2.py
```
In the tests/app1.py file I have put in these codes to direct pytest to look into the tests/app1/ folder for test. Repeat this for app2 so you can write different test for each different app.
```
    args = sys.argv[1:]
    if len(args) == 0:
        args = ["tests/app1"]
```
Then you can run briefcase to test your app with the following command
```
briefcase run --test --app app1
briefcase run --test --app app2
```

**Reply by freakboy3742** [reply]
Yup - that's about what I'd expect to see in a multi-app setup.
